src/wolfssl_bio.rs

```diff
@@ -0,0 +1,252 @@
+use openssl::{
+    asn1::Asn1Time,
+    bn::{BigNum, MsbOption},
+    hash::MessageDigest,
+    pkey::{PKey, PKeyRef, Private},
+    version::version,
+    x509::{
+        extension::{BasicConstraints, KeyUsage, SubjectKeyIdentifier},
+        X509NameBuilder, X509Ref, X509,
+    },
+};
+
+use crate::agent::TLSVersion;
+use crate::error::Error;
+use crate::io::MemoryStream;
+use crate::openssl_binding::static_rsa_cert;
+use wolfssl_sys as wolf;
+
+/* Note: for writing this, I tried to mimic the openssl::bio module. I adapted the calls to the C functions
+from OpenSSL to WolfSSL but the internal calling conventions might differ so we may need to rework
+this. */
+
+/*******************
+ WolfSSL BIO
+********************/
+
+pub type BIO = wolf::WOLFSSL_BIO;
+#[allow(bad_style)]
+unsafe fn BIO_set_flags(bio: *mut BIO, flags: c_int) {
+    (*bio).flags = flags;
+}
+
+#[allow(bad_style)]
+unsafe fn BIO_get_data(bio: *mut BIO) -> *mut c_void {
+    (*bio).ptr
+}
+
+#[allow(bad_style)]
+unsafe fn BIO_set_data(bio: *mut BIO, data: *mut c_void) {
+    (*bio).ptr = data;
+}
+
+#[allow(bad_style)]
+unsafe fn BIO_set_num(bio: *mut BIO, num: c_int) {
+    (*bio).num = num;
+}
+
+unsafe fn BIO_set_init(bio: *mut BIO, num: c_int) {
+    wolf::wolfSSL_BIO_set_init(bio, num);
+}
+
+pub const BIO_TYPE_NONE: c_int = 0;
+pub const BIO_CTRL_EOF: c_int = 2;
+pub const BIO_CTRL_INFO: c_int = 3;
+pub const BIO_CTRL_FLUSH: c_int = 11;
+pub const BIO_CTRL_DGRAM_QUERY_MTU: c_int = 40;
+pub const BIO_C_SET_BUF_MEM_EOF_RETURN: c_int = 130;
+pub const BIO_FLAGS_READ: c_int = 0x01;
+pub const BIO_FLAGS_WRITE: c_int = 0x02;
+pub const BIO_FLAGS_IO_SPECIAL: c_int = 0x04;
+pub const BIO_FLAGS_RWS: c_int = BIO_FLAGS_READ | BIO_FLAGS_WRITE | BIO_FLAGS_IO_SPECIAL;
+pub const BIO_FLAGS_SHOULD_RETRY: c_int = 0x08;
+
+pub unsafe fn BIO_set_retry_read(b: *mut BIO) {
+    BIO_set_flags(b, BIO_FLAGS_READ | BIO_FLAGS_SHOULD_RETRY)
+}
+
+pub unsafe fn BIO_set_retry_write(b: *mut BIO) {
+    BIO_set_flags(b, BIO_FLAGS_WRITE | BIO_FLAGS_SHOULD_RETRY)
+}
+
+pub unsafe fn BIO_clear_retry_flags(b: *mut BIO) {
+    wolf::wolfSSL_BIO_clear_flags(b, BIO_FLAGS_RWS | BIO_FLAGS_SHOULD_RETRY)
+}
+
+#[allow(clippy::match_like_matches_macro)] // matches macro requires rust 1.42.0
+fn retriable_error(err: &io::Error) -> bool {
+    match err.kind() {
+        ErrorKind::WouldBlock | ErrorKind::NotConnected => true,
+        _ => false,
+    }
+}
+
+/*******************
+ WolfSSL BIO METHOD
+********************/
+// struct BIO_METHOD(*mut wolf::WOLFSSL_BIO_METHOD);
+/// Safe wrapper for BIO_METHOD
+pub struct BioMethod(*mut wolf::WOLFSSL_BIO_METHOD);
+
+pub struct StreamState<S> {
+    pub stream: S,
+    pub error: Option<io::Error>,
+    pub panic: Option<Box<dyn Any + Send>>,
+    pub dtls_mtu_size: c_long,
+}
+unsafe fn state<'a, S: 'a>(bio: *mut BIO) -> &'a mut StreamState<S> {
+    &mut *(BIO_get_data(bio) as *mut _)
+}
+
+unsafe extern "C" fn bwrite<S: Write>(
```

Comment:
Again, I heavily rely on the OpenSSL compatibility layer of WolfSSL and I hope that the internal conventions are the same.
Maybe it would have been simpler to link wolfssl to the the rust-openssl crate after all but I failed when I tried this as the latter needs a lot more of functions to be exposed than what we need here.

Reply:
Yes, such unsafe code is required sadly.
I think its easier to do it that way. Else we would also have to maintain WolfSSL support in rust-openssl. Or fork it and having troubles to keep it up-to-date.

Reply:
Are we actually using the compatibility layer here? Did you find a C example which uses the wolfssl API instead of the OpenSSL API?